pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+
+				mapObj, _, _ := unstructured.NestedSlice(
+					fedPod.Object,
+					ctlutil.SpecField, ctlutil.TemplateField, "spec", "containers",
+				)
+				image := mapObj[0].(map[string]interface{})["image"]
+				Expect(image).To(BeEquivalentTo(expectedImage))
+			})
+		})
+
+		When("on update the Kube API returns an error that is not a NotFound", func() {
+			BeforeEach(func() {
+				clientPatcher = &fakeClientWithUpdateError{"test error"}
+			})
+
+			It("should return an error and not create the resource", func() {
+				err := f.Distribute(resource, clusterNames...)
+				Expect(err).To(MatchError(HavePrefix("test error")))
+			})
+		})
+
+		When("the resource is not in the Kube API", func() {
+			It("should create the resource", func() {
+				// Sanity check
+				_, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).To(HaveOccurred())
+
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(fedPod).ToNot(BeNil())
+				Expect(fedPod).To(BeEquivalentTo(fedResource))
+			})
+		})
+
+		When("the resource needs to be distributed to specific clusters", func() {
+			BeforeEach(func() {
+				clusterNames = []string{"cluster1", "cluster2"}
+			})
+
+			It("should set the placement to be an explicit list of clusters", func() {
+				// This test implicitly tests that the nested hierarchy of fields exists
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				mapObj, found, err := getClustersField(fedPod)
+				fedClusterNames := unpackClusterNames(mapObj)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+				Expect(fedClusterNames).To(BeEquivalentTo(clusterNames))
+			})
+
+			It("the returned resource's placement should not contain an empty matcher", func() {
+				// This test implicitly tests that the nested hierarchy of fields exists
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				mapObj, found, err := getMatchLabelsField(fedPod)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).ToNot(BeTrue(), "Hierarchy of fields is present!")
+				Expect(mapObj).To(BeEmpty())
+			})
+		})
+
+		When("the resource needs to be distributed to specific clusters", func() {
+			It("the returned resource's placement should be an empty matcher", func() {
+				// This test implicitly tests that the nested hierarchy of fields exists
+				_ = f.Distribute(resource)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				mapObj, found, err := getMatchLabelsField(fedPod)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+				Expect(mapObj).To(BeEmpty())
```

Comment:
Similar to above wrt suggested var name changes...

Reply:
You are welcome to change them in future patches.